[Woo POS][Refunds] Create full refund

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

WOOMOB-1855

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR allows creating a full refund in Woo POS in a happy path refund scenario.

💡 The UI of the loading states, error, and success states is not in scope of this PR-it will be updated in the next iteration.

### Test Steps
<!-- Describe how to test your changes. Include only what’s needed for the reviewer to validate the behavior.
For new features, outline the main user flow or goal rather than every tap or click — the reviewer should be able to complete the flow naturally.
If applicable, mention key devices, scenarios, or edge cases to verify. -->
1. Create order in POS and try creating a full refund through the POS Orders screen. To test:
a) order containing a single item
b) order containing multiple items
c) order containing multiple quantities of the same item 
3. Verify that the network request and server response look as expected.
4. Verify the order is properly refunded in wp-admin.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
https://github.com/user-attachments/assets/83d82eae-37be-435e-84a0-b5b51143abaa

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Allow creating refund for single quantity items

samiuelson committed:
Allow creating refund for items of multiple quantity

samiuelson committed:
Clean up code: move logic to use case class

samiuelson committed:
Add unit tests

samiuelson committed:
Merge branch 'trunk' into woomob-1855-create-refund

samiuelson committed:
Update tests

samiuelson committed:
Merge branch 'trunk' into woomob-1855-create-refund

samiuelson committed:
Allow refunding multiple quantity

samiuelson committed:
Update tests

samiuelson committed:
Refactor onUIEvent method to simplify state handling in WooPosRefundViewModel

samiuelson committed:
Refactor onUIEvent method to handle dialog dismissal logic and prevent state changes during processing

samiuelson committed:
Extract calculation logic to external use case classes

samiuelson committed:
Handle empty item lists in refund calculations to prevent errors

samiuelson committed:
Merge branch 'trunk' into woomob-1855-create-refund

samiuelson committed:
Revert optional refundTotal and refundTax

samiuelson commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/refunds/RefundRequestItem.kt:
```diff
@@ -9,12 +9,12 @@ data class RefundRequestItem(
+    val refundTotal: BigDecimal? = null,
     @SerializedName("refund_tax")
-    val refundTax: List<RefundRequestTax>
+    val refundTax: List<RefundRequestTax>? = null
```
Reverted: 882b9168cc4a2a53eaff13dca9d264cb75b1a913

samiuelson committed:
Refactor dialog dismissal handling and UI event processing in WooPosRefundViewModel

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosRefundViewModel.kt:
```diff
@@ -99,23 +111,74 @@ class WooPosRefundViewModel @AssistedInject constructor(
+                true
+            }
         }
+    }
```
Refactored: 6874490f3434c7a5e169e4ca0366cd6fb20f5fd4

samiuelson committed:
Fix refund tax calculation to use BigDecimal comparison for accuracy

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosCalculateRefundTax.kt:
```diff
@@ -0,0 +1,43 @@
+        }
+
+        return if (refundQuantity.toFloat() == originalItem.quantity) {
+            originalItem.totalTax
```
Done: a1dc6681d7bc4e231fe19b393af4f7c4e048697f

samiuelson committed:
Prevent processing refunds when already in processing state

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosRefundViewModel.kt:
```diff
@@ -99,23 +111,74 @@ class WooPosRefundViewModel @AssistedInject constructor(
+                    WooPosRefundUIEvent.BackToReviewClicked ->
+                        _state.value = currentState.copy(step = WooPosRefundState.Content.RefundStep.ReviewRefund)
+                    WooPosRefundUIEvent.OnRefundConfirmed -> processRefund(currentState)
+                    WooPosRefundUIEvent.DialogDismissed -> Unit
+                }
+                true
+            }
         }
+    }
+
+    private fun processRefund(contentState: WooPosRefundState.Content) {
+        viewModelScope.launch {
+            _state.value = contentState.copy(step = WooPosRefundState.Content.RefundStep.Processing)
```
Improved: 2699b4c3c897eded5e5792a3d734375c8161fa64

samiuelson committed:
Add tests

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosRefundViewModel.kt:
```diff
@@ -99,23 +111,74 @@ class WooPosRefundViewModel @AssistedInject constructor(
         }
+    }
+
+    private fun processRefund(contentState: WooPosRefundState.Content) {
+        viewModelScope.launch {
+            _state.value = contentState.copy(step = WooPosRefundState.Content.RefundStep.Processing)
 
-        _state.value = currentState.copy(step = newStep)
+            val order = currentOrder ?: run {
+                _state.value = WooPosRefundState.RefundError(
+                    message = resourceProvider.getString(R.string.error_generic)
+                )
+                return@launch
+            }
+
+            val refundItems = groupRefundItems(contentState.refundableItems, order)
+
+            val result = refundStore.createItemsRefund(
+                site = selectedSite.get(),
+                orderId = contentState.orderId,
+                amount = contentState.total,
+                reason = "",
+                restockItems = true,
+                autoRefund = false,
+                items = refundItems
+            )
+
+            if (result.isError) {
+                _state.value = WooPosRefundState.RefundError(
+                    message = result.error.message ?: resourceProvider.getString(R.string.error_generic)
+                )
+            } else {
+                _state.value = WooPosRefundState.RefundSuccess(
+                    orderId = contentState.orderId,
+                    orderNumber = contentState.orderNumber,
+                    refundedAmount = contentState.formattedTotal
+                )
+            }
+        }
     }
```
Added tests: 07097441ee9872cf4fb159769ca6eae6d936fc64

samiuelson committed:
Simplify refund subtotal calculation by removing unnecessary checks for empty item lists

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosGroupRefundItems.kt:
```diff
@@ -0,0 +1,62 @@
+    ): BigDecimal {
+        if (items.isEmpty()) {
+            return BigDecimal.ZERO
+        }
```
Simplified: bd8e0ac3a2ddbf26d06b6d6bcac622a7ef2175b1

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosCalculateRefundSubtotal.kt:
```diff
@@ -0,0 +1,20 @@
+                } else {
+                    val quantity = items.size.toBigDecimal()
+                    quantity.multiply(items.first().unitPrice)
+                }
```
Simplified: bd8e0ac3a2ddbf26d06b6d6bcac622a7ef2175b1

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosRefundViewModel.kt:
```diff
@@ -99,23 +111,74 @@ class WooPosRefundViewModel @AssistedInject constructor(
+                site = selectedSite.get(),
+                orderId = contentState.orderId,
+                amount = contentState.total,
+                reason = "",
```
Support for refund reason is not in scope of this PR.

samiuelson committed:
Prevent dialog dismissal during refund processing to ensure user sees the result

samiuelson committed:
Notify UI on dialog dismissal to manage refund state transitions

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosRefundViewModel.kt:
```diff
@@ -99,23 +111,74 @@ class WooPosRefundViewModel @AssistedInject constructor(
+                val currentState = _state.value
+                val isProcessing = currentState is WooPosRefundState.Content &&
+                    currentState.step == WooPosRefundState.Content.RefundStep.Processing
+
+                if (!isProcessing) {
+                    loadRefundableItems()
+                    true
+                } else {
+                    false
+                }
```
Improved in [8eb7f5b](https://github.com/woocommerce/woocommerce-android/pull/15135/commits/8eb7f5bdd33494f94c2a70451d0f1a70b0552b66) and [2f50bb6](https://github.com/woocommerce/woocommerce-android/pull/15135/commits/2f50bb64ac8906ce3fe0d08d7fdbbb3a47d830c0)

samiuelson committed:
Log error when currentOrder is null during refund processing

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosRefundViewModel.kt:
```diff
@@ -99,23 +111,74 @@ class WooPosRefundViewModel @AssistedInject constructor(
+    }
+
+    private fun processRefund(contentState: WooPosRefundState.Content) {
+        viewModelScope.launch {
+            _state.value = contentState.copy(step = WooPosRefundState.Content.RefundStep.Processing)
 
-        _state.value = currentState.copy(step = newStep)
+            val order = currentOrder ?: run {
+                _state.value = WooPosRefundState.RefundError(
+                    message = resourceProvider.getString(R.string.error_generic)
+                )
+                return@launch
+            }
```
Added log [6e9c2ed](https://github.com/woocommerce/woocommerce-android/pull/15135/commits/6e9c2ed6b70e0e2d66249d9fd4ee5584567c4183)

samiuelson committed:
Merge branch 'trunk' into woomob-1855-create-refund

samiuelson committed:
Merge branch 'trunk' into woomob-1855-create-refund

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Merge remote-tracking branch 'origin/woomob-1855-create-refund' into woomob-1855-create-refund

samiuelson committed:
Merge branch 'trunk' into woomob-1855-create-refund

samiuelson committed:
Rename function

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosCalculateRefundTax.kt:
```diff
@@ -0,0 +1,43 @@
+        return if (refundQuantity.toBigDecimal().compareTo(originalItem.quantity.toBigDecimal()) == 0) {
+            originalItem.totalTax
+        } else {
+            val singleItemTax = originalItem.totalTax.divide(
```
💡 This logic is an equivalent of `fun ProductRefundListItem.calculateTotalTaxes()` in store management. We are discussing improving the API to get rid of local calculations like these, but for now we decided to move forward with existing API that requires local calculations.

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosCalculateRefundTax.kt:
```diff
@@ -0,0 +1,43 @@
+    ): BigDecimal {
+        val originalItem = order.items.find { it.itemId == orderItemId }
+
+        if (originalItem == null || originalItem.quantity == 0f) {
```
❓ Is there a valid scenario in which originalItem is expected to be null or have quantity 0? If there isn't I believe we should crash here - money is involved and if we enter an unexpected state I believe it's safer to crash. Wdyt?

samiuelson replied:
This makes sense. Added exception throwing in case of invalid values: 49d81509ff46b180065d236e7a7e40215150e110

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosGroupRefundItems.kt:
```diff
@@ -0,0 +1,59 @@
+                RefundRequestItem(
+                    itemId = orderItemId,
+                    quantity = refundQuantity,
+                    refundTotal = calculateRefundTotal(items, refundQuantity),
```
❓ Should we pass `originalItem` here instead of `items`? I might be missing something, but I don't understand why we pass `items` for Totals calculation but `originalItem` for Taxes calculation.

samiuelson replied:
Good catch. Improved in: 3e68ba241f30b2f1205cf150746257d66dc2805a

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosRefundViewModel.kt:
```diff
@@ -99,23 +111,87 @@ class WooPosRefundViewModel @AssistedInject constructor(
 
-        _state.value = currentState.copy(step = newStep)
+    private fun processRefund(contentState: WooPosRefundState.Content) {
+        if (contentState.step == WooPosRefundState.Content.RefundStep.Processing) {
```
❓ I'm wondering if we need to do this check inside the coroutine, otherwise if we invoke process refund twice we might end up with a race condition:
1. Invoke ProcessRefund
2. Check state is not processing
3. Launch coroutine A
4. Invoke process refund second time
5. Check state is not processing
6. Launch coroutine B
7. Coroutine A starts
8. Coroutine B starts


Alternatively, we could invoke the `_state.value = contentState.copy(step = WooPosRefundState.Content.RefundStep.Processing)` before we launch the coroutine.

samiuelson replied:
Good catch. Fixed in: 9666f01

malinajirka reviewed (commented):
Thanks @samiuelson! 

I found some potential issues. Some of them are likely not bugs, but simply parts of the feature that aren't implemented yet.

1. The refund total was one cent off. There is a chance the same happens even in store management or even wpadmin, I haven't tested that.
<img width="350" height="263" alt="Screenshot 2026-01-06 at 11 55 05" src="https://github.com/user-attachments/assets/04280853-7247-4639-8973-db4ad024e31a" />
<img width="350" height="534" alt="Screenshot 2026-01-06 at 11 55 14" src="https://github.com/user-attachments/assets/1ca240e7-9650-4296-b150-64011d72ad5d" />
<img width="350" height="372" alt="Screenshot 2026-01-06 at 11 56 41" src="https://github.com/user-attachments/assets/b523d598-6ff2-4da9-a08b-3aa5b39debc7" />
<img width="350" height="418" alt="Screenshot 2026-01-06 at 12 02 37" src="https://github.com/user-attachments/assets/01164a72-d3a2-49d3-8fcf-27f63f1234bc" />


I was able to replicate this on another order on my store:
<img width="350" height="650" alt="Screenshot 2026-01-06 at 12 15 51" src="https://github.com/user-attachments/assets/67da47cd-9c7a-4f04-8e65-236945d00c4e" />


2. It seems even when I do a full refund, the order remains in `Completed` status instead of `Refunded`.

3. `Edit Reason` button doesn't do anything.

4. I noticed there is a cross/close icon on the 'processing refund' dialog - it doesn't do anything.

5. The refund VM seems to survive so when it's in a certain state (failed/success) I can't get it out of it:

https://github.com/user-attachments/assets/1300d482-89cc-47da-85c2-e36dccffbeb4

6. Some orders simply say 'Invalid refund amount`:
<img width="350" height="736" alt="Screenshot 2026-01-06 at 12 10 23" src="https://github.com/user-attachments/assets/cb5b7bbd-75af-4519-86d9-14f6c2b53fe0" />

Request
```
{
  "path": "/wc/v3/orders/5413/refunds/\u0026_method\u003dpost",
  "json": "true",
  "body": "{\"reason\":\"\",\"amount\":\"1.311835\",\"api_refund\":\"false\",\"line_items\":[{\"id\":6892,\"quantity\":1,\"refund_tax\":[{\"refund_total\":0.11,\"id\":3}],\"refund_total\":1.201835}],\"restock_items\":true}"
}
```
Response
```
{
  "error": "woocommerce_rest_cannot_create_order_refund",
  "message": "Invalid refund amount."
}
```

samiuelson committed:
Merge branch 'trunk' into woomob-1855-create-refund

samiuelson committed:
Enhance refund calculation by adding validation for order item existence and quantity

samiuelson committed:
Enhance WooPosGroupRefundItems

* Add validation for order items in refund calculations
* Use same original Order.Item for both taxes and refund total calculation

samiuelson committed:
Improve refund processing logic to prevent duplicate processing (race conditions)

samiuelson committed:
Use number of decimals parameter in refundable items calculation

samiuelson committed:
Add number of decimals parameter for refund calculations and update tests

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Merge branch 'trunk' into woomob-1855-create-refund

samiuelson committed:
Add number of decimals parameter to refund tax calculations

samiuelson committed:
Merge remote-tracking branch 'origin/woomob-1855-create-refund' into woomob-1855-create-refund

samiuelson committed:
Merge branch 'trunk' into woomob-1855-create-refund

samiuelson committed:
Fix typo

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/orders/WooPosGroupRefundItems.kt:
```diff
@@ -0,0 +1,76 @@
+        }
+
+        val refundQuantity = quantity.toBigDecimal()
+        val numberOfDecimals = wooCommerceStore.getSiteSettings(selectedSite.get())?.currencyDecimalNumber
```
🔍 Is it intentional that we retrieve the decimal number directly from wooCommerceStore here but in WooPosGroupRefundItems we pass it as a parameter from the VM?

samiuelson replied:
Cleaned up with: dba96931f6b1ce60bd29f96dddf80b0f09bc180d

malinajirka reviewed (commented):
Thanks for the updates @samiuelson! They look good and some of the issues are gone. However, I'm still able to reproduce the 'one cent off' issue (it more or less works in the store management part of the app). See the attached video with voice description of what's happening. 

https://github.com/user-attachments/assets/181551e0-a0a1-44b4-8e80-0ab398ef61e8


The order detail is not refreshed on successful refund (even refund which doesn't end up in the one cent off issue). This is reproduced and mentioned in the video as well.

I'm reproducing this on https://site-for-woocommerce12a3fasdf45dfs6789.mystagingwebsite.com/ with three Apple Ring products in the order. You are admin on the site. Let me know if I can help somehow.

samiuelson committed:
Apply rounding to total not unit values

samiuelson committed:
Update tests

samiuelson commented:
Thanks for review, @jirka! 

> The fix works for the Apple Ring, however, I'm still able to reproduce similar rounding issues. If I recall correctly, there is a setting in Core somewhere which says whether items should be rounded per unit vs per row in the order. It's a random guess, but it could perhaps be related?
> 
> You should be able to reproduce these issues on my store except of the first one(https://site-for-woocommerce12a3fasdf45dfs6789.mystagingwebsite.com/):
> 
> This one might not be easy to reproduce - I think it might be related to the fact that the missing item has been trashed/deleted from the store.

I checked the scenarios you shared, and as I mentioned internally, I can see that we are not able to handle them internally with the existing API. 

The root problem is related to the `Yes, I will enter prices inclusive of tax setting`. When enabled, both the product price and tax values often end up having multiple decimal places, and WooCommerce doesn't round them to any specific number of decimal points. 

* The `dp` request parameter affects only the number of decimal places the `price` is sent with (purely for presentation purposes, not for calculations). 

* The tax `total` value is always sent with **two** decimal places by the endpoint. 

However, internally WooCommerce backend uses unrounded values of price and tax total for calculations, in order to provide high precision. Since we don't have a precise enough value of the tax total, we are not able to replicate local calculations of the backend in the scenarios you provided. This leads to one-cent-off difference in the refund `amount` value–in case the refund `amount` is 0.01 lower the refund creation will succeed because the amount is lower than order total. In case it's higher the endpoint returns `Invalid refund amount` error. 

* This problem is also present in the store management refund flow.

cc @toupper

samiuelson committed:
Add decimal points parameter to: fetchOrders, fetchSingleOrder, fetchOrdersFromRemote and refreshOrderById

samiuelson committed:
Fix rounding logic in refund calculations

- Apply rounding at the final step of the refund total calculation to prevent intermediate rounding errors.
- Refresh order details after a refund is successfully processed to ensure the UI displays the most up-to-date information.

samiuelson committed:
Clean up code

Satisfy detekt's complaints

samiuelson committed:
Clean up code

samiuelson committed:
Clean up tests

samiuelson committed:
Clean up code

samiuelson committed:
Merge branch 'trunk' into woomob-1855-create-refund

malinajirka reviewed (approved):
Thanks for not giving up on this, it works much better now. 👏  I tested tens of different orders (on handle-taxes-rounding-setting branch), in different settings and all except one worked. I personally believe, this is good enough - especially since we plan to replace this solution with a server driven calculation.

The order which ended up with a wrong amount was Order#5509 on my store with `Round tax at subtotal level, instead of rounding per line` **unchecked**. What surprised me a bit is that when I re-create this order, no matter the above setting, the total is always 129.75 so I'm not sure where is the .76 coming from. 

<img height="400" alt="Image" src="https://github.com/user-attachments/assets/af9c89c5-3652-4897-b330-762cdb276e74" />
